File: desktop/Windows/native/paths.py
"""Все пути приложения в одном месте — чтобы было видно, где что лежит.

Во время разработки храним всё рядом с проектом (папки bin/ и data/),
а не в скрытых системных каталогах. Так проще проверить, что внутри.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path


def project_root() -> Path:
    """Корень проекта (папка, где лежит run.py).

    Работает и при запуске из исходников, и в собранном .exe (PyInstaller).
    """
    if getattr(sys, "frozen", False):  # собранный exe
        return Path(sys.executable).resolve().parent
    return Path(__file__).resolve().parent.parent


ROOT = project_root()

FROZEN = bool(getattr(sys, "frozen", False))

# Папка с ядром и гео-базами (xray.exe, geoip.dat, geosite.dat, sing-box, wintun).
# В обоих режимах — рядом с приложением (в собранном виде установщик кладёт сюда
# бинарники; они только читаются).
BIN_DIR = ROOT / "bin"


def _data_dir() -> Path:
    """Папка для пользовательских данных (профили, настройки, логи).

    В режиме разработки — рядом с проектом (data/), чтобы всё было под рукой.
    В собранном приложении — в %LOCALAPPDATA%\\SCVPN, т.к. рядом с exe в
    Program Files писать нельзя.
    """
    if FROZEN:
        base = os.environ.get("LOCALAPPDATA") or str(Path.home())
        return Path(base) / "SCVPN"
    return ROOT / "data"


# Папка с пользовательскими данными: профили, настройки, логи.
DATA_DIR = _data_dir()
LOG_DIR = DATA_DIR / "logs"

# Конкретные файлы.
PROFILES_FILE = DATA_DIR / "profiles.json"      # серверы + подписки
SETTINGS_FILE = DATA_DIR / "settings.json"      # настройки приложения
XRAY_CONFIG_FILE = DATA_DIR / "xray_running.json"  # активный конфиг ядра (для отладки)
# Активный .conf AmneziaWG. Лежит там же, где конфиг Xray, и защищён так же —
# то есть правами на саму папку: в собранном приложении это %LOCALAPPDATA%\SCVPN
# внутри профиля пользователя, в разработке — data/ рядом с исходниками. Отдельной
# защиты приватному ключу не выдумываем: рядом уже лежит xray_running.json с
# uuid и паролем сервера, и он ничем не менее ценен.
AWG_CONFIG_FILE = DATA_DIR / "awg_running.conf"


def xray_exe() -> Path:
    return BIN_DIR / ("xray.exe" if os.name == "nt" else "xray")


def awg_exe() -> Path:
    """scvpn-awg — AmneziaWG в виде локального SOCKS5 (сборка из папки awg/)."""
    return BIN_DIR / ("scvpn-awg.exe" if os.name == "nt" else "scvpn-awg")


def singbox_exe() -> Path:
    """sing-box используется только для TUN-режима (поднимает TUN-адаптер)."""
    return BIN_DIR / ("sing-box.exe" if os.name == "nt" else "sing-box")


def wintun_dll() -> Path:
    """Драйвер виртуального адаптера для TUN на Windows."""
    return BIN_DIR / "wintun.dll"


def geoip_dat() -> Path:
    return BIN_DIR / "geoip.dat"


def geosite_dat() -> Path:
    return BIN_DIR / "geosite.dat"


def resource_path(rel: str) -> Path:
    """Путь к упакованному ресурсу (иконка и т.п.).

    В собранном приложении PyInstaller распаковывает данные в _MEIPASS.
    """
    if FROZEN:
        base = Path(getattr(sys, "_MEIPASS", ROOT))
        return base / rel
    return ROOT / rel


def icon_file() -> Path:
    return resource_path("scvpn.ico") if FROZEN else (ROOT / "setup" / "scvpn.ico")


def ensure_dirs() -> None:
    """Создать рабочие папки, если их ещё нет."""
    for d in (BIN_DIR, DATA_DIR, LOG_DIR):
        d.mkdir(parents=True, exist_ok=True)

File: desktop/Windows-cs/SCVPN/Native/Paths.cs
using System;
using System.IO;

namespace SCVPN.Native;

/// <summary>
/// Все пути приложения в одном месте — чтобы было видно, где что лежит.
/// </summary>
/// <remarks>
/// Отличие от Python-версии: там была вилка «запуск из исходников / собранный
/// exe» (<c>sys.frozen</c>), и в режиме разработки данные лежали в <c>data/</c>
/// рядом с проектом. В .NET такой вилки нет — запуск всегда «собранный», —
/// поэтому режим разработки заменён переменной окружения
/// <c>SCVPN_DATA_DIR</c>: она же позволяет тестам увести запись во временную
/// папку, а не в настоящие профили пользователя (в Swift ради этого поля
/// сделаны изменяемыми, см. core-swift/Paths.swift).
///
/// Свойства считаются при каждом обращении, а не один раз при инициализации
/// типа: иначе подмена <c>SCVPN_DATA_DIR</c> из теста сработала бы только
/// если тест успел выставить её раньше первого чужого обращения к Paths.
///
/// Имена файлов на диске трогать нельзя — это файлы пользователя, оставшиеся
/// от Python-версии, и обновление обязано их подхватить.
/// </remarks>
public static class Paths
{
    /// <summary>Папка, из которой запущено приложение (рядом лежит exe).</summary>
    /// <remarks>
    /// <see cref="AppContext.BaseDirectory"/> оканчивается разделителем;
    /// срезаем его, чтобы путь можно было печатать в лог без «\\» в середине.
    /// Для склейки всё равно используется <see cref="Path.Combine(string, string)"/>.
    /// </remarks>
    public static string Root { get; } =
        Path.TrimEndingDirectorySeparator(AppContext.BaseDirectory);

    /// <summary>
    /// Папка с ядром и гео-базами (xray.exe, sing-box.exe, wintun.dll,
    /// geoip.dat, geosite.dat).
    /// </summary>
    /// <remarks>
    /// Рядом с приложением: установщик кладёт сюда бинарники, а приложение их
    /// только читает и перекачивает по кнопке. В %LOCALAPPDATA% их уносить
    /// нельзя — sing-box грузит wintun.dll из своей рабочей папки.
    /// </remarks>
    public static string BinDir => Path.Combine(Root, "bin");

    /// <summary>Папка с пользовательскими данными: профили, настройки, логи.</summary>
    /// <remarks>
    /// Рядом с exe писать нельзя: приложение стоит в Program Files. Если задана
    /// <c>SCVPN_DATA_DIR</c> — берём её (замена «режима разработки» из Python и
    /// точка подмены для тестов). Относительный путь в переменной считается от
    /// <see cref="Root"/>, а не от текущей папки процесса: у запущенного из
    /// ярлыка приложения текущая папка непредсказуема.
    /// </remarks>
    public static string DataDir
    {
        get
        {
            var custom = Environment.GetEnvironmentVariable("SCVPN_DATA_DIR");
            if (!string.IsNullOrWhiteSpace(custom))
            {
                // Path.Combine возвращает второй аргумент как есть, если он
                // абсолютный, — то есть эта строка покрывает оба случая.
                return Path.Combine(Root, custom.Trim());
            }

            var local = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
            if (string.IsNullOrEmpty(local))
            {
                // Тот же запасной путь, что в Python: профиль пользователя.
                local = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            }

            return Path.Combine(local, "SCVPN");
        }
    }

    public static string LogDir => Path.Combine(DataDir, "logs");

    /// <summary>Серверы и подписки.</summary>
    public static string ProfilesFile => Path.Combine(DataDir, "profiles.json");

    /// <summary>Настройки приложения.</summary>
    public static string SettingsFile => Path.Combine(DataDir, "settings.json");

    /// <summary>Активный конфиг Xray — лежит на диске ради отладки.</summary>
    public static string XrayConfigFile => Path.Combine(DataDir, "xray_running.json");

    /// <summary>Активный конфиг sing-box (TUN-режим) — тоже ради отладки.</summary>
    public static string SingboxConfigFile => Path.Combine(DataDir, "singbox_running.json");

    // Расширение .exe зашито: TFM у проекта Windows-only, ветка «не nt» из
    // Python здесь недостижима.
    public static string XrayExe => Path.Combine(BinDir, "xray.exe");

    /// <summary>sing-box нужен только для TUN-режима: он поднимает TUN-адаптер.</summary>
    public static string SingboxExe => Path.Combine(BinDir, "sing-box.exe");

    /// <summary>Драйвер виртуального адаптера для TUN на Windows.</summary>
    public static string WintunDll => Path.Combine(BinDir, "wintun.dll");

    public static string GeoipDat => Path.Combine(BinDir, "geoip.dat");

    public static string GeositeDat => Path.Combine(BinDir, "geosite.dat");

    /// <summary>Архитектура бинарников, лежащих сейчас в <see cref="BinDir"/>.</summary>
    /// <remarks>
    /// Без метки нельзя отличить arm64-набор от x64: пользователь, обновившийся
    /// на ARM-машине, остался бы со старыми x64-ядрами и получил невнятную
    /// ошибку запуска вместо предложения скачать ядро (см. шаг 3 плана
    /// docs/windows-arm64-plan.md). Отсутствие файла означает «x64»: так вёл
    /// себя bin/, собранный до появления метки.
    /// </remarks>
    public static string ArchStamp => Path.Combine(BinDir, "arch.txt");

    /// <summary>Иконка приложения файлом на диске.</summary>
    /// <remarks>
    /// Окну иконка достаётся ресурсом сборки (pack-URI), но части Windows API
    /// нужен именно файл. Файла может не оказаться — вызывающий обязан
    /// проверить существование, как это делал run_app() в Python.
    /// </remarks>
    public static string IconFile => Path.Combine(Root, "Assets", "scvpn.ico");

    /// <summary>Файл с PID запущенного ядра: <c>xray.pid</c> или <c>singbox.pid</c>.</summary>
    /// <remarks>
    /// Оба ядра — отдельные процессы и переживают аварийное закрытие
    /// приложения: туннель поднят, а управлять им уже нечем. Следующий запуск
    /// читает эти файлы и снимает осиротевшие процессы (Tun.CleanupStray).
    /// </remarks>
    public static string PidFile(string name) => Path.Combine(DataDir, name);

    /// <summary>Создать рабочие папки, если их ещё нет.</summary>
    public static void EnsureDirs()
    {
        // bin/ лежит в Program Files: у обычного пользователя прав на создание
        // может не быть. Падать из-за этого на старте нельзя — про нехватку
        // прав скажет тот код, который реально туда пишет (скачивание ядра),
        // и скажет понятнее. Так же поступает Swift-версия (try?).
        foreach (var dir in new[] { BinDir, DataDir, LogDir })
        {
            try
            {
                Directory.CreateDirectory(dir);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
            }
        }
    }
}
